File: src/p0002/golang/solution.go
func addTwoNumbers(a *ListNode, b *ListNode) *ListNode {
	var (
		sentinel ListNode
		tail     = &sentinel
	)

	for carry := 0; carry != 0 || a != nil || b != nil; carry /= 10 {
		if a != nil {
			carry, a = carry+a.Val, a.Next
		}

		if b != nil {
			carry, b = carry+b.Val, b.Next
		}

		tail.Next = &ListNode{Val: carry % 10}
		tail = tail.Next
	}

	return sentinel.Next
}

File: src/p0002/python/solution.py
class Solution:
    @staticmethod
    def addTwoNumbers(a: Optional[ListNode], b: Optional[ListNode]) -> Optional[ListNode]:
        sentinel = ListNode(0)
        tail = sentinel
        carry = 0
        while carry or a or b:
            if a:
                carry, a = carry + a.val, a.next
            if b:
                carry, b = carry + b.val, b.next
            tail.next = ListNode(carry % 10)
            tail = tail.next
            carry = carry // 10
        return sentinel.next
